lib/paginator/ecto/query/dynamic_filter_builder.ex

```diff
@@ -0,0 +1,29 @@
+defmodule Paginator.Ecto.Query.DynamicFilterBuilder do
+  @type dynamic_filter_builder :: (integer(), term(), Ecto.Query.t() -> Ecto.Query.t())
+
+  @callback build_dynamic_filter(:after | :before, term()) :: dynamic_filter_builder()
+
+  @dispatch_table %{
+    desc: Paginator.Ecto.Query.DescNullsFirst,
+    desc_nulls_first: Paginator.Ecto.Query.DescNullsFirst,
+    desc_nulls_last: Paginator.Ecto.Query.DescNullsLast,
+    asc: Paginator.Ecto.Query.AscNullsLast,
```

Comment:
this is using Postgres defaults. Other databases may behave differently, but I'm not exactly sure how to add support for that in the library. What do you think?

Reply:
I was considering defining a specific `:asc/:desc` implementation that throws on null values. The problem is that if the database sorts nulls values differently, the pagination will be incorrect.

Reply:
I'm not sure if I have any good ideas here, or if I've understood the problem correctly 🤔 

The only thought (a bit half baked admittedly 😅 ) is that we could perhaps define a protocol which would then allow 3rd parties to define their own implementations for different databases by defining structs and implementations? Something roughly like the below:

```elixir
defmodule Paginator.Ecto.Query.DynamicFilterBuilder do
  @type dynamic_filter_builder :: (integer(), term(), Ecto.Query.t() -> Ecto.Query.t())

  @callback build_dynamic_filter(:after | :before, term()) :: dynamic_filter_builder()

  defprotocol DynamicFilter do
    @spec build!(t, :after | :before, term()) ::
            (integer(), term(), Ecto.Query.t() -> Ecto.Query.t())
    def build!(sort_order, direction, value)
  end

  defimpl DynamicFilter, for: Atom do
    @dispatch_table %{
      desc: Paginator.Ecto.Query.DescNullsFirst,
      desc_nulls_first: Paginator.Ecto.Query.DescNullsFirst,
      desc_nulls_last: Paginator.Ecto.Query.DescNullsLast,
      asc: Paginator.Ecto.Query.AscNullsLast,
      asc_nulls_last: Paginator.Ecto.Query.AscNullsLast,
      asc_nulls_first: Paginator.Ecto.Query.AscNullsFirst
    }

    def build!(sort_order, direction, value) do
      case Map.fetch(@dispatch_table, sort_order) do
        {:ok, module} ->
          apply(module, :build_dynamic_filter, [direction, value])

        :error ->
          available_sort_orders = Map.keys(@dispatch_table) |> Enum.join(", ")

          raise(
            "Invalid sorting value :#{sort_order}, please please use either #{available_sort_orders}"
          )
      end
    end
  end

  # 3-rd party could define their own struct to use when sorting
  defmodule MysqlDynamicFilter do
    defstruct [:sort_order]
  end

  # and then define an implementation for how that sorting should behave
  defimpl DynamicFilter, for: MysqlDynamicFilter do
    def build!(%{sort_order: _sort_order}, _direction, _value) do
      # create mysql specific filter clauses
    end
  end

  @spec builder!(DynamicFilter.t(), :after | :before, term()) :: dynamic_filter_builder()
  def builder!(sort_order, direction, value) do
    DynamicFilter.build!(sort_order, direction, value)
  end
end
```

Again, not sure if I've fully understood the problem we are trying to solve so apologies if this is complete gibberish 😆

Reply:
No worries, thanks a lot for the suggestion. I don't think that we need a different implementation because the current implementation should cover all cases. I think that what we really need is some mechanism to define which implementation is used when `asc` and `desc` are given.

The crux of the problem is that mysql does not support the extra variants (nulls asc/desc) and the defaults are different from postgres. In Mysql, the following applies:

```
asc = asc-nulls-first
desc = desc-nulls-last
```

I'm wondering if perhaps using the adapter, instead of only the `sort_order` could work here:

```
Postgrex => %{
  asc: AscNullsLast,
  desc: DescNullsFirst,
},

Mysql => %{
  asc: AscNullsFirst,
  desc: DescNullsLast,
}
```

what do you think?

Reply:
On second thought, the adapter is not available in this location (I think it is better to address this issue in a follow up PR, if that is Ok).

Reply:
Ah I see, thanks for the explanation.

> On second thought, the adapter is not available in this location. I think it is better to address this issue in a follow up PR, if that is Ok.

Agreed, sounds good to me